CONTRIBUTING.md

```diff
@@ -142,3 +125,59 @@ We have some of this in the Reducer code, but generally discourage it.
+
+**Use immutable types when it doesn't hurt the performance**
+Wrap object types in [Readonly](https://www.typescriptlang.org/docs/handbook/utility-types.html#readonlytype) or mark individual fields as `readonly`.
+
+**Don't use namespaces**
+Use native ES modules instead, as [recommended by TypeScript documentation](https://www.typescriptlang.org/docs/handbook/namespaces-and-modules.html#using-modules).
+
+**Avoid `any` as much as possible**
+It's almost always possible to type things properly with modern Typescript.
+
+**Always use `===` instead of `==`**
+Loose equality is [crazy](https://dorey.github.io/JavaScript-Equality-Table/unified/).
+
+**Don't use too many external libraries**
+Heuristics for deciding whether pulling an external library is worth it:
+
+- is it maintained? is it going to stay maintained in the future?
+- how hard it would be to reimplement the functionality ourselves? if it's a few lines, it's better to control the implementation in our own codebase
+- what's the bundle size of the dependency and would it be effectively tree-shaked, if the dependency is big and we need only a small part of it?
+- would we want to fine-tune the implementation in the future, because of Squiggle's design needs or for the sake of performance?
+  - the closer to the "core" functionality of Squiggle the feature is (math, distributions-related code), the more it makes sense to keep control over the implementation details
+
+**Prefer named exports over default exports**
+It's easier to do refactorings with named exports.
+
+**Name files according to their main named exports; split code into many small files**
+This is expecially straightforward in the frontend code; try to put one component in a single file, `export const MyComponent: React.FC = ...` from `MyComponent.tsx`.
+
+In the squiggle-lang code, I'm not sure yet if this is viable.
+
+**Prefer `type` over `interface`**
+In the modern TypeScript there's no [big](https://stackoverflow.com/questions/37233735/interfaces-vs-types-in-typescript/52682220) [difference](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#differences-between-type-aliases-and-interfaces) between types and interfaces. Errors in interfaces can be slightly nicer, but interfaces are open by default and we mostly don't want that.
+
+It's not worth fighting over, though, the difference is pretty small.
+
+**Prefer arrow functions over `function` keyword whenever possible**
+Arrow functions have less hidden features (`this`, `arguments`, function hoisting) and are more concise when writing in functional style.
+
+It's ok to use `function` when you need to define a complicated generic function, e.g. a generic functional React component.
+
+**Prefer `undefined` over `null`**
+
+**Use `UpperCamelCase` for type names, `camelCase` for variable names**
+
+**Use exceptions instead of Ok/Error pairs, wrap in try/catch on top level if necessary**
```

Comment:
I imagine that later, we could use a lot more thinking in the best patterns for exceptions/error handling